optimizer/lib/transformers/BlurImagePlaceholders.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const sizeOf = require('image-size');
+const jimp = require('jimp');
+
+/**
+ * Adds placeholders for all AMP images and posters that are blurry versions of
+ * the corresponding original source. The blur will be displayed as the
+ * <amp-img> is rendering, and will fade out once the element is loaded.
+ */
+
+class BlurImagePlaceholders {
+  /**
+   * Parses the document to add blurred placedholders in all appropriate
+   * locations.
+   * @param {Tree} tree - a DOM tree.
+   * @param {Object} params - a dictionary containing transformer specific parameters.
+   */
+  transform(tree, params) {
+    const html = tree.root.firstChildByTag('html');
+    const body = html.firstChildByTag('body');
+    const promises = [];
+    for (let node = body; node !== null; node = node.nextNode()) {
+      let childSrc;
+      if (node.tagName === 'amp-img') {
+        childSrc = node.attribs.src;
+      }
+      if (node.tagName === 'amp-video' && node.attribs.poster) {
+        childSrc = node.attribs.poster;
+      }
+      if (childSrc) {
+        let parent = node;
```

Comment:
Does `parent` need to exist?

Reply:
So back when I used a while loop instead of the for loop, I needed to store parent because the pointer associated with node was updating before the appending of the child image. The work around with this was by storing the appropriate node as a parent. Maybe the while loop was the problem, so I'll try it without it now!

Reply:
Originally when I was using a while loop to traverse the tree, the appending of the placeholder would not work because the reference to node would be different than intended due to the time it took for all the placeholder code to execute. The solution for this was to store a reference to the correct node in parent. 

Now with the for loop, the tests seem to be passing while removing the "parent" reference, but I'm not really sure why because I thought that the code should still be executing synchronously like with the while loop. Is there a difference in the for vs the while loop that makes it work now?